users/zeineldeen/experiments/conformer_att_2022/librispeech_960/seq_train_helpers.py

```diff
@@ -168,3 +168,79 @@ def add_min_wer(net, loss_scale, ce_scale, am_scale, lm_scale, beam_size, ce_lab
         },
         "target": "bpe_labels",
     }
+
+
+def add_mmi(net, loss_scale, ce_scale, am_scale, lm_scale, beam_size, ce_label_smoothing=0.0):
+    """
+    Add Maximum Mutual Information training criterion.
+    """
+    assert ce_label_smoothing==0.0, "label smoothing not implemented correctly"
+    subnet = net["output"]["unit"]
+
+    # add LM
+    subnet["lm_output"] = get_trans_lstm_lm()
+    subnet["lm_output_prob"] = {
+        "class": "activation",
+        "activation": "softmax",
+        "from": ["lm_output"],
+        "target": "bpe_labels",
+    }  # [B,V]
+
+    # add extra search subnet to return the N-best list
+    net["extra.search:output"] = copy.deepcopy(net["output"])
+    extra_search_subnet = net["extra.search:output"]["unit"]
+
+    extra_search_subnet["output_prob"].pop("loss", None)
+    extra_search_subnet["output_prob"].pop("loss_opts", None)
+    extra_search_subnet["output_prob"].pop("loss_scale", None)
+
+    extra_search_subnet["asr_output_log_prob"] = {
+        "class": "activation",
+        "activation": "safe_log",
+        "from": "output_prob",
+    }
+    extra_search_subnet["scaled_asr_output_log_prob"] = {
+        "class": "eval",
+        "from": "asr_output_log_prob",
+        "eval": f"source(0) * {am_scale}",
+    }
+    extra_search_subnet["lm_output_log_prob"] = {
+        "class": "activation",
+        "activation": "safe_log",
+        "from": "lm_output_prob",
+    }
+    extra_search_subnet["scaled_lm_output_log_prob"] = {
+        "class": "eval",
+        "from": "lm_output_log_prob",
+        "eval": f"source(0) * {lm_scale}",
+    }
+    extra_search_subnet["combined_log_prob"] = {
+        "class": "combine",
+        "kind": "add",
+        "from": ["scaled_asr_output_log_prob", "scaled_lm_output_log_prob"],
+    }
+    extra_search_subnet["output"]["from"] = "combined_log_prob"  # asr + lm
+    extra_search_subnet["output"]["length_normalization"] = False
+    extra_search_subnet["output"]["cheating"] = "exclusive"
+    extra_search_subnet["output"]["beam_size"] = beam_size
+    extra_search_subnet["output"]["input_type"] = "log_prob"
+
+    # CE loss
+    if "loss_opts" in subnet["output_prob"]:
+        subnet["output_prob"]["loss_opts"]["scale"] = loss_scale + ce_scale
```

Comment:
so this is correct because the AM part of the MMI numerator is same as CE loss right?

Reply:
Correct, cf https://arxiv.org/abs/2005.10049 Equation 7.
The `loss_scale` is for the numerator part (with the \alpha)
The `ce_scale` is if some additional cross entropy loss should be added.

Note that because of this being part of the numerator, I require `ce_label_smoothing==0.0`, as this would otherwise interfere with MMI loss. we could duplicate this output layer and then have the MMI numerator without label smoothing and the CE with optional label smoothing.